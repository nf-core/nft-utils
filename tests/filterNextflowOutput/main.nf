include { FASTQC } from './modules/local/fastqc'

params.failure = false
params.outdir  = "results"
workflow {
  input = channel.of(
    'sample_1',
    'sample_2',
    'sample_3',
  )

  FASTQC(input)

  // Create a file in the output directory

  // Write to files for validation
  def output_dir = params.outdir

  // Create output directory
  new File(output_dir).mkdirs()

  // Write citation text
  new File("${output_dir}/test.txt").text = "This is a test file"

  def test_file = file("${output_dir}/test.txt")

  log.info(
    """
\033[2m----------------------------------------------------\033[0m-
                                        \033[0;32m,--.\033[0;30m/\033[0;32m,-.\033[0m
\033[0;34m        ___     __   __   __   ___     \033[0;32m/,-._.--~\'\033[0m
\033[0;34m  |\\ | |__  __ /  ` /  \\ |__) |__         \033[0;33m}  {\033[0m
\033[0;34m  | \\| |       \\__, \\__/ |  \\ |___     \033[0;32m\\`-._,-`-,\033[0m
                                        \033[0;32m`._,._,\'\033[0m

\033[0;37m      ____\033[0m
\033[0;37m    .´ _  `.\033[0m
\033[0;37m   /  \033[0;32m|\\\033[0m`-_ \\\033[0m     \033[0;34m __        __   ___     \033[0m
\033[0;37m  |   \033[0;32m| \\\033[0m  `-|\033[0m    \033[0;34m|__`  /\\  |__) |__  |__/\033[0m
\033[0;37m   \\ \033[0;32m|   \\\033[0m  /\033[0m     \033[0;34m.__| /¯¯\\ |  \\ |___ |  \\\033[0m
\033[0;37m    `\033[0;32m|\033[0m____\033[0;32m\\\033[0m´\033[0m
-\033[2m----------------------------------------------------\033[0m-
nf-core/pipeline1: 1.0dev
nf-core/pipeline2: 1.0
nf-core/pipeline3: 1.1.0dev
nf-core/pipeline4: 1.1.0
        """
  )

  log.info("outdir: ${params.outdir}")
  log.info("HOME: ${System.getenv("HOME")}")
  log.info("NFT_WORKDIR: ${System.getenv("NFT_WORKDIR")}")
  log.info("NXF_CACHE_DIR: ${System.getenv("NXF_CACHE_DIR")}")
  log.info("NXF_CONDA_CACHEDIR: ${System.getenv("NXF_CONDA_CACHEDIR")}")
  log.info("NXF_HOME: ${System.getenv("NXF_HOME")}")
  log.info("NXF_SINGULARITY_CACHEDIR: ${System.getenv("NXF_SINGULARITY_CACHEDIR")}")
  log.info("NXF_TEMP: ${System.getenv("NXF_TEMP")}")
  log.info("NXF_WORK: ${System.getenv("NXF_WORK")}")

  println("println message")
  log.info("\033[0;34mlog.info message\033[0m")
  log.warn("\033[0;33mlog.warn message\033[0m")
  log.error("\033[0;31mlog.error message\033[0m")

  println("println with path: ${test_file}")
  log.info("\033[0;34mlog.info with path: ${test_file}\033[0m")
  log.warn("\033[0;33mlog.warn with path: ${test_file}\033[0m")
  log.error("\033[0;31mlog.error with path: ${test_file}\033[0m")
  if (params.failure) {
    System.err.println("\033[1;31mSystem error with path: ${test_file}\033[0m")
    System.err.println("\033[1;31mSystem error message\033[0m")
    error("\033[1;31mError with path: ${test_file}\033[0m")
  }
}
