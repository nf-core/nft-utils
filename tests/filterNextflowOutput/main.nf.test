nextflow_pipeline {

    name "Test filterNextflowOutput"
    script "./main.nf"
    tag "filterNextflowOutput"

    test("Capture stdout, stderr and both when failure") {

        tag 'both'
        tag 'failure'
        tag 'stderr'
        tag 'stdout'

        when {
            params {
                failure = true
            }
        }

        then {
            def filtered_stdout = filterNextflowOutput(workflow.stdout)
            def filtered_stderr = filterNextflowOutput(workflow.stderr)

            // Print the stdout
            println "workflow.stdout [${workflow.stdout.size()}]:"
            println workflow.stdout.join('\n')

            // Print the filtered stdout
            println "filtered_stdout: [${filtered_stdout.size()}]"
            println filtered_stdout.join('\n')

            // Print the stderr
            println "workflow.stderr [${workflow.stderr.size()}]:"
            println workflow.stderr.join('\n')

            // Print the filtered stderr
            println "filtered_stderr: [${filtered_stderr.size()}]"
            println filtered_stderr.join('\n')

            assert workflow.failed
            assert snapshot(
                filterNextflowOutput(workflow.stdout),
                filterNextflowOutput(workflow.stderr, include: ["System error message"])?:"no stderr",
                filterNextflowOutput(workflow.stdout + workflow.stderr)
              ).match()
        }
    }


    test("Capture both stdout and stderr and keep ANSI codes") {

        tag 'ansi'
        tag 'both'
        tag 'stderr'
        tag 'stdout'

        then {
            def filtered_stdout = filterNextflowOutput(workflow.stdout, keepAnsi: true)
            def filtered_stderr = filterNextflowOutput(workflow.stderr, keepAnsi: true)

            // Print the stdout
            println "workflow.stdout [${workflow.stdout.size()}]:"
            println workflow.stdout.join('\n')

            // Print the filtered stdout
            println "filtered_stdout: [${filtered_stdout.size()}]"
            println filtered_stdout.join('\n')

            // Print the stderr
            println "workflow.stderr [${workflow.stderr.size()}]:"
            println workflow.stderr.join('\n')

            // Print the filtered stderr
            println "filtered_stderr: [${filtered_stderr.size()}]"
            println filtered_stderr.join('\n')

            assert workflow.success
            assert snapshot(
                filterNextflowOutput(workflow.stdout, keepAnsi: true, ignore: ["Submitted process"]),
                filterNextflowOutput(workflow.stderr, keepAnsi: true)?:"no stderr",
                filterNextflowOutput(workflow.stdout + workflow.stderr, keepAnsi: true)
            ).match()
        }
    }

    test("Do not sort") {

        tag 'sorted'

        then {
            def filtered_stdout = filterNextflowOutput(workflow.stdout, sorted: false)

            // Print the stdout
            println "workflow.stdout [${workflow.stdout.size()}]:"
            println workflow.stdout.join('\n')

            // Print the filtered stdout
            println "filtered_stdout: [${filtered_stdout.size()}]"
            println filtered_stdout.join('\n')

            assert workflow.success
        }
    }
}
