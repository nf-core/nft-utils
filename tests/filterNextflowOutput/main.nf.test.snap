{
    "Capture stdout, stderr and both when failure": {
        "content": [
            [
                "N E X T F L O W  ~  version 25.04.6",
                "Launching `[PATH]/tests/filterNextflowOutput/./main.nf` [RUN_NAME] DSL2 - revision: [REVISION]",
                "\u001b[2m----------------------------------------------------\u001b[0m-",
                "                                        \u001b[0;32m,--.\u001b[0;30m/\u001b[0;32m,-.\u001b[0m",
                "\u001b[0;34m        ___     __   __   __   ___     \u001b[0;32m/,-._.--~'\u001b[0m",
                "\u001b[0;34m  |\\ | |__  __ /  ` /  \\ |__) |__         \u001b[0;33m}  {\u001b[0m",
                "\u001b[0;34m  | \\| |       \\__, \\__/ |  \\ |___     \u001b[0;32m\\`-._,-`-,\u001b[0m",
                "                                        \u001b[0;32m`._,._,'\u001b[0m",
                "\u001b[0;37m      ____\u001b[0m",
                "\u001b[0;37m    .\u00b4 _  `.\u001b[0m",
                "\u001b[0;37m   /  \u001b[0;32m|\\\u001b[0m`-_ \\\u001b[0m     \u001b[0;34m __        __   ___     \u001b[0m",
                "\u001b[0;37m  |   \u001b[0;32m| \\\u001b[0m  `-|\u001b[0m    \u001b[0;34m|__`  /\\  |__) |__  |__/\u001b[0m",
                "\u001b[0;37m   \\ \u001b[0;32m|   \\\u001b[0m  /\u001b[0m     \u001b[0;34m.__| /\u00af\u00af\\ |  \\ |___ |  \\\u001b[0m",
                "\u001b[0;37m    `\u001b[0;32m|\u001b[0m____\u001b[0;32m\\\u001b[0m\u00b4\u001b[0m",
                "-\u001b[2m----------------------------------------------------\u001b[0m-",
                "nf-core/pipeline1: [VERSION]",
                "nf-core/pipeline2: [VERSION]",
                "nf-core/pipeline3: [VERSION]",
                "nf-core/pipeline4: [VERSION]",
                "outdir: results",
                "HOME: [PATH]",
                "NFT_WORKDIR: [PATH]",
                "NXF_CACHE_DIR: null",
                "NXF_CONDA_CACHEDIR: null",
                "NXF_HOME: [PATH]",
                "NXF_SINGULARITY_CACHEDIR: [PATH]",
                "NXF_TEMP: null",
                "NXF_WORK: null",
                "println message",
                "\u001b[0;34mlog.info message\u001b[0m",
                "println with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt",
                "\u001b[0;34mlog.info with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "\u001b[1;31mError with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                " -- Check '[PATH]/.nf-test/tests/[NFT_HASH]/meta/nextflow.log' file for details",
                "ERROR ~ \u001b[0;31mlog.error message\u001b[0m",
                "ERROR ~ \u001b[0;31mlog.error with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "WARN: \u001b[0;33mlog.warn message\u001b[0m",
                "WARN: \u001b[0;33mlog.warn with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m"
            ],
            [
                "\u001b[1;31mSystem error with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "\u001b[1;31mSystem error message\u001b[0m"
            ],
            [
                "N E X T F L O W  ~  version 25.04.6",
                "Launching `[PATH]/tests/filterNextflowOutput/./main.nf` [RUN_NAME] DSL2 - revision: [REVISION]",
                "\u001b[2m----------------------------------------------------\u001b[0m-",
                "                                        \u001b[0;32m,--.\u001b[0;30m/\u001b[0;32m,-.\u001b[0m",
                "\u001b[0;34m        ___     __   __   __   ___     \u001b[0;32m/,-._.--~'\u001b[0m",
                "\u001b[0;34m  |\\ | |__  __ /  ` /  \\ |__) |__         \u001b[0;33m}  {\u001b[0m",
                "\u001b[0;34m  | \\| |       \\__, \\__/ |  \\ |___     \u001b[0;32m\\`-._,-`-,\u001b[0m",
                "                                        \u001b[0;32m`._,._,'\u001b[0m",
                "\u001b[0;37m      ____\u001b[0m",
                "\u001b[0;37m    .\u00b4 _  `.\u001b[0m",
                "\u001b[0;37m   /  \u001b[0;32m|\\\u001b[0m`-_ \\\u001b[0m     \u001b[0;34m __        __   ___     \u001b[0m",
                "\u001b[0;37m  |   \u001b[0;32m| \\\u001b[0m  `-|\u001b[0m    \u001b[0;34m|__`  /\\  |__) |__  |__/\u001b[0m",
                "\u001b[0;37m   \\ \u001b[0;32m|   \\\u001b[0m  /\u001b[0m     \u001b[0;34m.__| /\u00af\u00af\\ |  \\ |___ |  \\\u001b[0m",
                "\u001b[0;37m    `\u001b[0;32m|\u001b[0m____\u001b[0;32m\\\u001b[0m\u00b4\u001b[0m",
                "-\u001b[2m----------------------------------------------------\u001b[0m-",
                "nf-core/pipeline1: [VERSION]",
                "nf-core/pipeline2: [VERSION]",
                "nf-core/pipeline3: [VERSION]",
                "nf-core/pipeline4: [VERSION]",
                "outdir: results",
                "HOME: [PATH]",
                "NFT_WORKDIR: [PATH]",
                "NXF_CACHE_DIR: null",
                "NXF_CONDA_CACHEDIR: null",
                "NXF_HOME: [PATH]",
                "NXF_SINGULARITY_CACHEDIR: [PATH]",
                "NXF_TEMP: null",
                "NXF_WORK: null",
                "println message",
                "\u001b[0;34mlog.info message\u001b[0m",
                "println with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt",
                "\u001b[0;34mlog.info with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "\u001b[1;31mError with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "\u001b[1;31mSystem error with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "\u001b[1;31mSystem error message\u001b[0m",
                " -- Check '[PATH]/.nf-test/tests/[NFT_HASH]/meta/nextflow.log' file for details",
                "ERROR ~ \u001b[0;31mlog.error message\u001b[0m",
                "ERROR ~ \u001b[0;31mlog.error with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "WARN: \u001b[0;33mlog.warn message\u001b[0m",
                "WARN: \u001b[0;33mlog.warn with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-03T13:07:47.431325897"
    },
    "Capture stderr only": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-03T12:08:37.687964135"
    },
    "Capture both stdout and stderr": {
        "content": [
            [
                "N E X T F L O W  ~  version 25.04.6",
                "Launching `[PATH]/tests/filterNextflowOutput/./main.nf` [RUN_NAME] DSL2 - revision: [REVISION]",
                "\u001b[2m----------------------------------------------------\u001b[0m-",
                "                                        \u001b[0;32m,--.\u001b[0;30m/\u001b[0;32m,-.\u001b[0m",
                "\u001b[0;34m        ___     __   __   __   ___     \u001b[0;32m/,-._.--~'\u001b[0m",
                "\u001b[0;34m  |\\ | |__  __ /  ` /  \\ |__) |__         \u001b[0;33m}  {\u001b[0m",
                "\u001b[0;34m  | \\| |       \\__, \\__/ |  \\ |___     \u001b[0;32m\\`-._,-`-,\u001b[0m",
                "                                        \u001b[0;32m`._,._,'\u001b[0m",
                "\u001b[0;37m      ____\u001b[0m",
                "\u001b[0;37m    .\u00b4 _  `.\u001b[0m",
                "\u001b[0;37m   /  \u001b[0;32m|\\\u001b[0m`-_ \\\u001b[0m     \u001b[0;34m __        __   ___     \u001b[0m",
                "\u001b[0;37m  |   \u001b[0;32m| \\\u001b[0m  `-|\u001b[0m    \u001b[0;34m|__`  /\\  |__) |__  |__/\u001b[0m",
                "\u001b[0;37m   \\ \u001b[0;32m|   \\\u001b[0m  /\u001b[0m     \u001b[0;34m.__| /\u00af\u00af\\ |  \\ |___ |  \\\u001b[0m",
                "\u001b[0;37m    `\u001b[0;32m|\u001b[0m____\u001b[0;32m\\\u001b[0m\u00b4\u001b[0m",
                "-\u001b[2m----------------------------------------------------\u001b[0m-",
                "nf-core/pipeline1: [VERSION]",
                "nf-core/pipeline2: [VERSION]",
                "nf-core/pipeline3: [VERSION]",
                "nf-core/pipeline4: [VERSION]",
                "outdir: results",
                "HOME: [PATH]",
                "NFT_WORKDIR: [PATH]",
                "NXF_CACHE_DIR: null",
                "NXF_CONDA_CACHEDIR: null",
                "NXF_HOME: [PATH]",
                "NXF_SINGULARITY_CACHEDIR: [PATH]",
                "NXF_TEMP: null",
                "NXF_WORK: null",
                "println message",
                "\u001b[0;34mlog.info message\u001b[0m",
                "println with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt",
                "\u001b[0;34mlog.info with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                " -- Check '[PATH]/.nf-test/tests/[NFT_HASH]/meta/nextflow.log' file for details",
                "ERROR ~ \u001b[0;31mlog.error message\u001b[0m",
                "ERROR ~ \u001b[0;31mlog.error with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "WARN: \u001b[0;33mlog.warn message\u001b[0m",
                "WARN: \u001b[0;33mlog.warn with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "[NXF_HASH] Submitted process > FASTQC (sample_1)",
                "[NXF_HASH] Submitted process > FASTQC (sample_2)",
                "[NXF_HASH] Submitted process > FASTQC (sample_3)"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-03T13:07:43.839607395"
    },
    "Capture stdout only": {
        "content": [
            [
                "N E X T F L O W  ~  version 25.04.6",
                "Launching `[PATH]/tests/filterNextflowOutput/./main.nf` [RUN_NAME] DSL2 - revision: [REVISION]",
                "\u001b[2m----------------------------------------------------\u001b[0m-",
                "                                        \u001b[0;32m,--.\u001b[0;30m/\u001b[0;32m,-.\u001b[0m",
                "\u001b[0;34m        ___     __   __   __   ___     \u001b[0;32m/,-._.--~'\u001b[0m",
                "\u001b[0;34m  |\\ | |__  __ /  ` /  \\ |__) |__         \u001b[0;33m}  {\u001b[0m",
                "\u001b[0;34m  | \\| |       \\__, \\__/ |  \\ |___     \u001b[0;32m\\`-._,-`-,\u001b[0m",
                "                                        \u001b[0;32m`._,._,'\u001b[0m",
                "\u001b[0;37m      ____\u001b[0m",
                "\u001b[0;37m    .\u00b4 _  `.\u001b[0m",
                "\u001b[0;37m   /  \u001b[0;32m|\\\u001b[0m`-_ \\\u001b[0m     \u001b[0;34m __        __   ___     \u001b[0m",
                "\u001b[0;37m  |   \u001b[0;32m| \\\u001b[0m  `-|\u001b[0m    \u001b[0;34m|__`  /\\  |__) |__  |__/\u001b[0m",
                "\u001b[0;37m   \\ \u001b[0;32m|   \\\u001b[0m  /\u001b[0m     \u001b[0;34m.__| /\u00af\u00af\\ |  \\ |___ |  \\\u001b[0m",
                "\u001b[0;37m    `\u001b[0;32m|\u001b[0m____\u001b[0;32m\\\u001b[0m\u00b4\u001b[0m",
                "-\u001b[2m----------------------------------------------------\u001b[0m-",
                "nf-core/pipeline1: [VERSION]",
                "nf-core/pipeline2: [VERSION]",
                "nf-core/pipeline3: [VERSION]",
                "nf-core/pipeline4: [VERSION]",
                "outdir: results",
                "HOME: [PATH]",
                "NFT_WORKDIR: [PATH]",
                "NXF_CACHE_DIR: null",
                "NXF_CONDA_CACHEDIR: null",
                "NXF_HOME: [PATH]",
                "NXF_SINGULARITY_CACHEDIR: [PATH]",
                "NXF_TEMP: null",
                "NXF_WORK: null",
                "println message",
                "\u001b[0;34mlog.info message\u001b[0m",
                "println with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt",
                "\u001b[0;34mlog.info with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                " -- Check '[PATH]/.nf-test/tests/[NFT_HASH]/meta/nextflow.log' file for details",
                "ERROR ~ \u001b[0;31mlog.error message\u001b[0m",
                "ERROR ~ \u001b[0;31mlog.error with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "WARN: \u001b[0;33mlog.warn message\u001b[0m",
                "WARN: \u001b[0;33mlog.warn with path: [PATH]/.nf-test/tests/[NFT_HASH]/results/test.txt\u001b[0m",
                "[NXF_HASH] Submitted process > FASTQC (sample_1)",
                "[NXF_HASH] Submitted process > FASTQC (sample_2)",
                "[NXF_HASH] Submitted process > FASTQC (sample_3)"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-03T13:07:35.368885788"
    }
}